types/src/executable.rs

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{access_path::AccessPath, state_store::state_key::StateKey};
+use aptos_crypto::HashValue;
+use std::sync::Arc;
+
+#[derive(PartialEq)]
+pub enum ExecutableDescriptor {
+    /// Possibly speculative, based on code published during the block.
+    Published(HashValue),
+
+    /// Based on code published (and committed) in previous blocks.
+    Storage,
+}
+
+pub trait ModulePath {
+    fn module_path(&self) -> Option<AccessPath>;
+}
+
+impl ModulePath for StateKey {
+    fn module_path(&self) -> Option<AccessPath> {
+        if let StateKey::AccessPath(ap) = self {
+            if ap.is_code() {
+                return Some(ap.clone());
+            }
+        }
+        None
+    }
+}
+
+/// For now we will handle the VM code cache / arena memory consumption on the
+/// executor side, likely naively in the beginning (e.g. flushing after a threshold).
+/// For the executor to manage memory consumption, executables should provide size.
+/// Note: explore finer-grained eviction mechanisms, e.g. LRU-based, or having
+/// different ownership for the arena / memory.
+pub trait Executable {
+    fn size(&self) -> usize;
+}
+
+pub struct ExecutableTestType(());
+
+impl Executable for ExecutableTestType {
+    fn size(&self) -> usize {
+        0
+    }
+}
+
+pub enum FetchedModule<X: Executable> {
+    Blob(anyhow::Result<Option<Vec<u8>>>),
+    // TODO: compiled module when available to avoid deserialization.
+    Executable(Arc<X>),
+}
+
+/// View for the VM for interacting with the multi-versioned executable cache.
+pub trait ExecutableView {
+    type Key;
+    type Executable: Executable;
+
+    /// This is an optimization to bypass transactional semantics and share the
```

Comment:
Can you elaborate more on the plan here? What are we bypassing? we will call MVHasmMap anyway, right? Are we going to have a separate data view and code view that both will call MVHashMap?

Reply:
When we produce an executable based on some module (that isn't from the running transaction), we let MVHashMap know immediately instead of waiting for the transaction to finish.